sources/engine/Xenko.Graphics/UIBatch.cs

```diff
@@ -517,28 +518,32 @@ protected override unsafe void UpdateBufferValuesFromElementInfo(ref ElementInfo
 
         private static unsafe void CalculateCubeVertices(UIImageDrawInfo* drawInfo, VertexPositionColorTextureSwizzle* vertex)
         {
+            const int VertexCountPerAxis = 2;
+
+            var depthBiasMultiplier = drawInfo->DepthBias * DepthBiasShiftOneUnit;
+            var colorScale = drawInfo->ColorScale.ToColor4();
+            var colorAdd = drawInfo->ColorAdd.ToColor4();
+            var swizzle = (float)drawInfo->Swizzle;
+
+            // The vertices are arranged from the top-left-back, going left to right first, then top to bottom, then back to front
             var currentPosition = drawInfo->LeftTopCornerWorld;
-            
-            // set the two first line of vertices
-            for (var l = 0; l < 2; ++l)
+            for (int l = 0; l < VertexCountPerAxis; l++)
             {
-                for (var r = 0; r < 2; r++)
+                for (int r = 0; r < VertexCountPerAxis; r++)
                 {
-                    for (var c = 0; c < 2; c++)
+                    for (var c = 0; c < VertexCountPerAxis; c++)
                     {
-                        vertex->ColorScale = drawInfo->ColorScale;
-                        vertex->ColorAdd = drawInfo->ColorAdd;
-
-                        vertex->Swizzle = (int)drawInfo->Swizzle;
-                        vertex->TextureCoordinate.X = 0; // cubes are used only for color
-                        vertex->TextureCoordinate.Y = 0; // cubes are used only for color
+                        vertex->ColorScale = colorScale;
+                        vertex->ColorAdd = colorAdd;
+                        vertex->Swizzle = swizzle;
```

Comment:
Looks like the value was clamped to an `int` before, but not anymore. Was it on purpose?

Reply:
To be honest, the swizzle lines confused me a little.
`drawInfo.Swizzle` is an enum (on both UIBatch & Sprite3DBatch), whereas the ones on the vertex data is a float, so there is a final conversion to float.
I'd like to think casting enum to int doesn't do anything, but starting to doubt myself.
I can convert it to int first if needed.

Reply:
I only reviewed it by sight :wink:. I haven't used Xenko for quite a while, so I'll let the experts figure this out.

Good job trying to improve the performance in any case.